contracts/ERC721A.sol

```diff
@@ -107,74 +150,106 @@ contract ERC721A is IERC721A {
      */
     function _getAux(address owner) internal view returns (uint64) {
-        return _addressData[owner].aux;
+        return uint64(_packedAddressData[owner] >> BITPOS_AUX);
     }
 
     /**
      * Sets the auxillary data for `owner`. (e.g. number of whitelist mint slots used).
      * If there are multiple variables, please pack them into a uint64.
      */
     function _setAux(address owner, uint64 aux) internal {
-        _addressData[owner].aux = aux;
+        uint256 packed = _packedAddressData[owner];
+        packed = (packed & BITMASK_AUX_COMPLEMENT) | (uint256(aux) << BITPOS_AUX);
+        _packedAddressData[owner] = packed;
     }
 
     /**
-     * Gas spent here starts off proportional to the maximum mint batch size.
-     * It gradually moves to O(1) as tokens get transferred around in the collection over time.
+     * Returns the packed ownership data of `tokenId`.
      */
-    function _ownershipOf(uint256 tokenId) internal view returns (TokenOwnership memory) {
+    function _packedOwnershipOf(uint256 tokenId) private view returns (uint256) {
         uint256 curr = tokenId;
 
         unchecked {
             if (_startTokenId() <= curr) if (curr < _currentIndex) {
-                TokenOwnership memory ownership = _ownerships[curr];
-                if (!ownership.burned) {
-                    if (ownership.addr != address(0)) {
-                        return ownership;
-                    }
+                uint256 packed = _packedOwnerships[curr];
+                // If not burned.
+                if (packed & BITMASK_BURNED == 0) {
                     // Invariant:
                     // There will always be an ownership that has an address and is not burned
                     // before an ownership that does not have an address and is not burned.
                     // Hence, curr will not underflow.
-                    while (true) {
-                        curr--;
-                        ownership = _ownerships[curr];
-                        if (ownership.addr != address(0)) {
-                            return ownership;
-                        }
+                    //
+                    // We can directly compare the packed value.
+                    // If the address is zero, packed is zero.
+                    while (packed == 0) {
+                        packed = _packedOwnerships[--curr];
                     }
+                    return packed;
                 }
             }
         }
         revert OwnerQueryForNonexistentToken();
     }
 
+    /**
+     * Returns the unpacked `TokenOwnership` struct from `packed`.
+     */
+    function _unpackedOwnership(uint256 packed) private pure returns (TokenOwnership memory ownership) {
+        ownership.addr = address(uint160(packed));
+        ownership.startTimestamp = uint64(packed >> BITPOS_START_TIMESTAMP);
+        ownership.burned = packed & BITMASK_BURNED != 0;
```

Comment:
Directly masking will be better than including a function. 

Returning boolean can have overhead.